### PR TITLE
feat: New tool to Google Search

### PR DESCRIPTION
**Pull request by nickscamara**
Related to #199 

Motivation: SerpAPI is very expensive and scrapping can cause problems.

Solution: Implemented the new Google search API done through the Programmable search engine.

It is a bit longer to set up but worth it as it gives 10,000 search queries per day for free.

Instructions:
    1. Install google-api-python-client
    - If you don't already have a Google account, sign up.
    - If you have never created a Google APIs Console project,
    read the Managing Projects page and create a project in the Google API Console.
    - Install the library using `pip install google-api-python-client`
    The current version of the library is 2.70.0 at this time

    2. To create an API key:
    - Navigate to the APIs & Services→Credentials panel in Cloud Console.
    - Select Create credentials, then select API key from the drop-down menu.
    - The API key created dialog box displays your newly created key.
    - You now have an `API_KEY`

    3. Setup Custom Search Engine so you can search the entire web
    - Create a custom search engine in this link.
    - In Sites to search, add any valid URL (i.e. www.stackoverflow.com).
    - That’s all you have to fill up, the rest doesn’t matter.
    In the left-side menu, click Edit search engine → {your search engine name}
    → Setup Set Search the entire web to ON. Remove the URL you added from
     the list of Sites to search.
    - Under Search engine ID you’ll find the `search-engine-ID`.

    4. Enable the Custom Search API
    - Navigate to the APIs & Services→Dashboard panel in Cloud Console.
    - Click Enable APIs and Services.
    - Search for Custom Search API and click on it.
    - Click Enable.
    URL for it: https://console.cloud.google.com/apis/library/customsearch.googleapis
    .com
    Adapted from: Instructions adapated from https://stackoverflow.com/questions/37083058/programmatically-searching-google-in-python-using-custom-search

- [X] Implementation
- [X] Tests, Inline Docs, Formatting
- [ ] Add it to `load_tools`.
- [ ] Improve external documentation

I think it still needs some general work here. Has been a while since I've coded in Python. I tried my best to follow the steps provided in the resources.

### CODE REVIEW

File: langchain/google_search.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
+"""Chain that calls Google Search."""
+import os
+import sys
+from typing import Any, Dict, List, Optional
+
+from googleapiclient.discovery import build
```
Comment (hwchase17) [inline langchain/google_search.py:6]: this should probably go inside the function (so its not a requirement for the package)

File: langchain/google_search.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+from pydantic import BaseModel, Extra, root_validator
+
+from langchain.utils import get_from_dict_or_env
+
+
+class HiddenPrints:
```
Comment (hwchase17) [inline langchain/google_search.py:12]: dont think this is used

File: langchain/google_search.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        )
+        values["google_api_key"] = google_api_key
+
+        google_cse_id = get_from_dict_or_env(values, "google_cse_id", "GOOGLE_CSE_ID")
+        values[google_cse_id] = google_cse_id
+
+        try:
+            from googleapiclient.discovery import build
+
+            build("customsearch", "v1", developerKey=google_api_key)
```
Comment (hwchase17) [inline langchain/google_search.py:101]: this is just to try connecting? could we pass this into init as search engine and use on line 83?

File: langchain/google_search.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        except Exception:
+            raise ValueError(
+                """Error in Google Search API, make sure you have
+                GOOGLE_API_KEY and GOOGLE_CSE_ID set on your enviroment.
+                If you have exceeded your 10,000/day token, please refer
+                to the google cse documentation."""
+            )
+
+
+# For backwards compatability
```
Comment (hwchase17) [inline langchain/google_search.py:134]: not needed!

Comment (nickscamara) [reply langchain/google_search.py]: Fixed!

File: langchain/google_search.py (hunk 3 above)
Comment (nickscamara) [reply langchain/google_search.py]: I think i fixed, but not sure if I did what you were thinking. Feel free to modify it

**Review by ilyazub — COMMENTED** [review]
@nickscamara thanks for your work!

> scrapping can cause problems.

SerpApi provides a scraper legal protection. I'm not a lawyer, but scraping public data is legal in the USA, the country in which SerpApi operates.

**Comment by jghaines** [comment]
Is this feature now released? I don't see any references to it in documentation.

**Pull request merged**